internal/provider/resource_workspace_run.go

```diff
@@ -0,0 +1,386 @@
+func NewWorkspaceRunResource() resource.Resource {
+	return &workspaceRunResource{}
+}
+
+type workspaceRunResource struct {
+	client *tharsis.Client
+}
+
+// Metadata returns the full name of the resource, including prefix, underscore, instance name.
+func (t *workspaceRunResource) Metadata(ctx context.Context,
+	req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = "tharsis_workspace_run"
+}
+
+func (t *workspaceRunResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	description := "Defines and manages a workspace run."
+
+	resp.Schema = schema.Schema{
+		Version:             1,
+		MarkdownDescription: description,
+		Description:         description,
+		Attributes: map[string]schema.Attribute{
+			"workspace_path": schema.StringAttribute{
+				MarkdownDescription: "The full path of the workspace.",
+				Description:         "The full path of the workspace.",
+				Required:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"module_source": schema.StringAttribute{
+				MarkdownDescription: "The source of the module, including the API hostname.",
+				Description:         "The source of the module, including the API hostname.",
+				Required:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"module_version": schema.StringAttribute{
+				MarkdownDescription: "The version identifier of the module.",
+				Description:         "The version identifier of the module.",
+				Optional:            true,
+				Computed:            true, // computed if not supplied
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"variables": schema.StringAttribute{
+				MarkdownDescription: "Optional variables for the run in the target workspace.",
+				Description:         "Optional variables for the run in the target workspace.",
+				Optional:            true,
+				// Will remain unset if not supplied.
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+		},
+	}
+}
+
+// Configure lets the provider implement the ResourceWithConfigure interface.
+func (t *workspaceRunResource) Configure(_ context.Context,
+	req resource.ConfigureRequest, _ *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	t.client = req.ProviderData.(*tharsis.Client)
+}
+
+func (t *workspaceRunResource) Create(ctx context.Context,
+	req resource.CreateRequest, resp *resource.CreateResponse) {
+
+	// Retrieve values from workspace run.
+	var workspaceRun WorkspaceRunModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &workspaceRun)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	created := t.doApplyOrDestroyRun(ctx, workspaceRun, false, resp.Diagnostics)
```

Comment:
This should only create a run if the current state wasn't deployed with the specified module/version.

Reply:
The only time an apply shouldn't happen is if there is no change to state regardless of the module and version.

Reply:
Actually this will need to take variables into account as well.